keps/sig-storage/3314-csi-differential-snapshot/README.md

```diff
@@ -0,0 +1,1061 @@
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Segmenting Critical Datapath Resource Requirements
+
+It's not uncommon that the response payload from the storage providers can be
+quite large, in the magnitude of gigabytes of changed block metadata, per volume
+snapshot pair. The usual declarative approach of updating the custom
+resource's `status` subresource with the payload may put etcd at risk of
+storage capacity exhaustion as well as regular IOPS spikes.
+
+To mitigate this issue, this KEP proposes a 2-hops request mechanism where
+instead of handling the requests directly, the `DifferentialSnapshot` controller
+returns a callback URL to the user so that HTTP requests can be directed to this
+"out-of-band" endpoint to fetch the changed block metadata. This mechanism
+allows the response payloads to be returned directly to the user, without
+persisting them in etcd.
+
+#### Securing The HTTP Listener
+
+The `DifferentialSnapshot` listener that exposes the callback endpoint will be
+secured by delegating the request authorisation to the Kubernetes API server
+using the `SubjectAccessReview` API.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The proposed design involves extending CSI with the `DifferentialSnapshot` CRD
+and the `DIFFERENTIAL_SNAPSHOT_SERVICE` capability. A Kubernetes user (human or
+not) would create a `DifferentialSnapshot` resource to initialize the CBT
+datapath workflow. Storage providers can opt in to support this feature by
+implementing the `DIFFERENTIAL_SNAPSHOT_SERVICE` capability in their CSI
+drivers.
+
+The `DifferentialSnapshot` resource is a namespace-scoped resource. It must be
+created in the same namespace as the base and target CSI `VolumeSnapshot`s.
+
+### CBT Datapath Worklow
+
+A Kubernetes user initiates the datapath workflow by creating a new
+`DifferentialSnapshot` custom resource. The `DifferentialSnapshot` controller,
+deployed as a CSI sidecar in the storage provider's CSI driver, watches for new
+`DifferentialSnapshot` resources:
+
+![CBT Step 1](./img/cbt-step-01.png)
+
+Instead of satisfying the request directly, the controller constructs a callback
+URL and adds it to the resource's `status` subresource. The user can then use
+the callback URL to fetch the list of changed block metadata:
+
+![CBT Step 2](./img/cbt-step-02.png)
+
+The callback URL points to the `DifferentialSnapshot` listener. The CSI driver
+`Service` resource's FQDN will be used as the hostname of the callback URL, which
```

Comment:
It has to be a `Service` that selects the CSI driver pods. Alternately, it can also be the [pod DNS](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#a-aaaa-records-1).